Intake and Ejector Configurator

Dhruv-0-Arora opened the pull request:
### Description
Implementation of configuring the positioning of a robot's intake and ejector. Uses raycasting to detect the node (very basic implementation of detecting node) and saves the information. Allows user to scale the size of the sphere that spawns as well.

### Features
- [x] Configure pickup zone
- [x] Configure ejector
- [x] Create physics sensor for pickup zone
- [x] Attach ejectable game piece
- [x] Eject ejectable game piece

### Robot with configured pickup zone
![image](https://github.com/user-attachments/assets/77c17bd7-223b-4a5d-bd74-619f163eb8bb)

### Robot holding a Game Piece
![image](https://github.com/user-attachments/assets/7eae0d3d-668f-4b2c-82f7-fb24b70bc310)

### Intake Configuration
![image](https://github.com/user-attachments/assets/12c6a25f-50cd-41e9-afd0-fa5c5e8d2639)

### Ejector Configuration
![image](https://github.com/user-attachments/assets/5fc74f84-0a61-4ca1-807d-9ef205fafa73)

### Temporary Components
These components are going to be left in and will be addressed shortly after in the scoring zone branch and likely one more branch to address input and behavior issues.

- Pieces are picked up via a ray cast from the intake zone, 3 meters forward along the z axis.
- Ejection is bound to left shift and control and isn't compatible with the InputSystem
- Pickup zones are always rendered
- There is no highlighting of rigid nodes

> [!Warning]
> Merge after #1007

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1728)

Dhruv-0-Arora committed:
Adding configure pickup in panelui
Need to merge transform gizmo code to add a sphere in the scene for implementation.

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dhruv/1733/mirabuf-gizmo' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Merge branch 'dhruv/1733/mirabuf-gizmo' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dhruv/1733/mirabuf-gizmo' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Added transform gizmo bug fixes and also cancel functionality

Dhruv-0-Arora committed:
Readded slider for sizing and configuration for the slider

Dhruv-0-Arora committed:
Merging dev

Dhruv-0-Arora committed:
Merge branch 'dhruv/1733/mirabuf-gizmo' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Quick fix
Removing duplicate line of code after merge

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/barclah/1682/mouse-interaction' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Merge branch 'dhruv/1733/mirabuf-gizmo' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Added scaling functionality with slider

Dhruv-0-Arora committed:
Raycasting work and basic implementation of pickup configuration

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Added storing data functionality

Dhruv-0-Arora committed:
Added configuration for shot trajectory

azaleacolburn commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -155,6 +207,12 @@ class MirabufSceneObject extends SceneObject {
         return mesh
     }
+
+    private EnablePhysics() {
+        this._mirabufInstance.parser.rigidNodes.forEach(rn => {
+            World.PhysicsSystem.EnablePhysicsForBody(this._mechanism.GetBodyByNodeId(rn.id)!)
+        })
+    }
```
Why is this a function, but disabling physics for all rigid bodies in the same way isn't? It's not an issue, I'm just curious.

Dhruv-0-Arora replied:
I added that to solve a bug so that is a good point and ill add that. The problem is that all of that stuff is from PR #1009 so I might need to wait for that to be merged

azaleacolburn commented on fission/src/systems/physics/PhysicsSystem.ts:
```diff
@@ -91,6 +92,26 @@ class PhysicsSystem extends WorldSystem {
+        this.GetBody(bodyId).SetIsSensor(true)
+    }
+
+    /**
+     * Enabing physics for a single body
+     *
+     * @param bodyId
+     */
+    public EnablePhysicsForBody(bodyId: Jolt.BodyID) {
+        this._joltBodyInterface.ActivateBody(bodyId)
+        this.GetBody(bodyId).SetIsSensor(false)
+    }
```
5 lines of comments, 4 lines of code. This is what programming's all about. To be honest, these functions are completely self-documenting without any comments.

Dhruv-0-Arora replied:
I don't even know why I did that but I do know its gone from #1009 😇

azaleacolburn commented on fission/src/systems/physics/PhysicsSystem.ts:
```diff
@@ -658,6 +682,91 @@ class PhysicsSystem extends WorldSystem {
+                        new JOLT.IndexedTriangle(indexArr.at(i)!, indexArr.at(i + 1)!, indexArr.at(i + 2)!, 0)
+                    )
+                }
+            }
```
Utilize the `?` operator to reduce redundancy

```suggestion
            const vertArr = body.triangleMesh?.mesh?.verts
            const indexArr = body.triangleMesh?.mesh?.indices
            if (vertArr && indexArr) {
                for (let i = 0; i < vertArr.length; i += 3) {
                    const vert = MirabufFloatArr_JoltFloat3(vertArr, i)
                    settings.mTriangleVertices.push_back(vert)
                    this.UpdateMinMaxBounds(new JOLT.Vec3(vert), min, max)
                    JOLT.destroy(vert)
                }
                for (let i = 0; i < indexArr.length; i += 3) {
                    settings.mIndexedTriangles.push_back(
                        new JOLT.IndexedTriangle(indexArr.at(i)!, indexArr.at(i + 1)!, indexArr.at(i + 2)!, 0)
                    )
                }
            }
```

azaleacolburn commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -198,57 +184,87 @@ class SceneRenderer extends WorldSystem {
-                    if (isDragging) {
-                        return
-                    }
+                if (!event.value && !Array.from(this.transformControls.keys()).some(tc => tc.dragging)) {
```
Extracting this operation to a variable might be nice

Dhruv-0-Arora replied:
That is true it would be more optimized

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/barclah/1682/mouse-interaction' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dhruv/1733/mirabuf-gizmo' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Bug fix + ran prettier + fixed transform gizmo size from merge

LucaHaverty committed:
Merge branch 'haverty/1692/preferences' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Prevented select node from selecting a field or the ground
need to fix bug of onCancel function not being run when another MainHUD panel being opened

LucaHaverty committed:
Ejector and intake config integration with preferences

LucaHaverty committed:
Merge branch 'dhruv/1728/intake-ejector-config' of github.com:autodesk/synthesis into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Ran prettier

Dhruv-0-Arora committed:
Created a working formula to calculate the position of the pickup object
need to optimize the code and add documentation

Dhruv-0-Arora committed:
Add documentation + fixed the fields showing in the select panel + optimized calculations

Dhruv-0-Arora committed:
Fixed filtering robots and fields + reimplemented the selectNode feature

Dhruv-0-Arora committed:
Optimized getting rotation and added implementation for configuring the shot trajectory
Rotating the shot trajectory rectangular prism is currently not functional

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora commented:
Notes for Reviewers
1. checkselectednode, listrobots, etc.. are duplicated functions. Where should I put them
2. Can I optimized this in checkSelectedNode()
``` ts
selectedRobot?.mirabufInstance?.parser.rigidNodes.forEach(rn => {
            if (World.PhysicsSystem.GetBody(selectedRobot.mechanism.GetBodyByNodeId(rn.id)!).GetID() === body.GetID()) {
                bodyAttachmentRef.current = body
                returnValue = true
            }
        })
```

HunterBarclay committed:
Solved gizmo transformation delta issue for ejector configuration.
Updated the ejector config gizmo to be a low poly cone, plus orientation
adjustments.
Fixed issue with ejector velocity saving constantly.
Refactoring most of ejector configuration.
Created pausing for the physics system.
Created physics body associations for associating additional data to
bodies.
Updated slider to more closely mimic MUI sliders.
Added proper node selection for ejector config.

HunterBarclay committed:
Updated intake preferences and copied ejector over to intake. Minor tweaks to adjust for scaling as well.

HunterBarclay committed:
Formatted

HunterBarclay committed:
Added new gamepiece association for detecting gamepieces

HunterBarclay committed:
Added intake sensor

HunterBarclay committed:
Ejectables done

HunterBarclay committed:
Adjusted max ejection velocity

HunterBarclay committed:
Merge branch 'haverty/1692/preferences' into dhruv/1728/intake-ejector-config

HunterBarclay committed:
Fixed label for intake size

HunterBarclay committed:
Formatted

HunterBarclay committed:
Merge branch 'haverty/1692/preferences' into dhruv/1728/intake-ejector-config

azaleacolburn commented on fission/src/mirabuf/EjectableSceneObject.ts:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    public Setup(): void {
+        if (this._parentAssembly.ejectorPreferences && this._gamePieceBodyId) {
```
With if statements like this one and the analogous one in Update, reversing the condition seems to be common practice to void code nesting. This isn't an issue here, but in other places (like Eject) you reverse the condition, so you might want to do the same here. Btw the reverse of `x && y` is `!x || !y`

HunterBarclay replied:
Generally speaking, I see doing `if (condition) { early out }` to be indicating that the function should have never been ran and it is a safe guard against poor user usage. Doing `if (condition) { then do thing }` indicates more that the function should've been ran, regardless of the state of condition. This also leaves it open for other things outside of the if statement to be ran potentially.

HunterBarclay replied:
And at the end of the day, it's up to the compiler and the operating system to determine how it performs the jump instruction and predictive statements, not which way you write the if statement.

azaleacolburn replied:
Oh ya, I didn't mean it was an optimization at all, I just meant we should be consistent with which choices we make. I like what you just said about what they indicate, that works for me 👍

azaleacolburn commented on fission/src/mirabuf/IntakeSensorSceneObject.ts:
```diff
@@ -0,0 +1,102 @@
+            World.PhysicsSystem.SetBodyPosition(this._joltBodyId, ThreeVector3_JoltVec3(position))
+            World.PhysicsSystem.SetBodyRotation(this._joltBodyId, ThreeQuaternion_JoltQuat(rotation))
+
+            if (!World.PhysicsSystem.isPaused) {
+                const hitRes = World.PhysicsSystem.RayCast(ThreeVector3_JoltVec3(position), new JOLT.Vec3(0, 0, 3))
+                if (hitRes) {
+                    const gpAssoc = World.PhysicsSystem.GetBodyAssociation<RigidNodeAssociate>(hitRes.data.mBodyID)
+                    if (gpAssoc && gpAssoc.isGamePiece) {
+                        console.debug("Found game piece!")
+                        this._parentAssembly.SetEjectable(hitRes.data.mBodyID, false)
+                    }
+                }
+            }
+
+            if (this._mesh) {
+                this._mesh.position.setFromMatrixPosition(bodyTransform)
+                this._mesh.rotation.setFromRotationMatrix(bodyTransform)
+            }
```
Not sure if this is possible, but you might be able to switch the position of these two if statements, then reverse each condition in the deeply nested one. If you can't switch them then nvm.

HunterBarclay replied:
I can switch the statements, but I'm not reversing the conditions as explained for the code in Ejectable

azaleacolburn replied:
Ya ignore this then 👍

azaleacolburn commented on fission/src/mirabuf/MirabufParser.ts:
```diff
@@ -60,8 +62,10 @@ class MirabufParser {
-    public get rigidNodes(): Array<RigidNodeReadOnly> {
-        return this._rigidNodes.map(x => new RigidNodeReadOnly(x))
+    public get rigidNodes(): Map<RigidNodeId, RigidNodeReadOnly> {
+        const result = new Map<RigidNodeId, RigidNodeReadOnly>()
+        this._rigidNodes.forEach(x => result.set(x.id, new RigidNodeReadOnly(x)))
```
You could use the [map higher order function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) instead to [convert an array into a map](https://stackoverflow.com/questions/75290595/convert-array-to-map-typescript) simplify these lines.

HunterBarclay replied:
I assumed that wasn't a thing for TypeScript, thanks.

azaleacolburn commented on fission/src/systems/physics/PhysicsSystem.ts:
```diff
@@ -90,6 +100,64 @@ class PhysicsSystem extends WorldSystem {
+        const res = this._bodyAssociations.get(bodyId.GetIndexAndSequenceNumber())
+        if (res) {
+            // Avoids error, simply returns undefined if invalid
+            return res as unknown as T
```
If you have to use unknown as a middleman, will this always be a safe cast given the class constraints on T?

HunterBarclay replied:
I'm fairly certain that this cast doesn't actually properly type check and I don't know how to fix it

HunterBarclay replied:
I originally had a separate BodyAssociated object for gamepieces, but literally not matter what I did, it would somehow successfully force cast a RigidNodeAssociate to GamePieceAssociate. I had required object types that weren't allowed to be null just fully forced to be null and I have no idea why. You can do normal `as` but then it throws an error on fail instead of just returning undefined.

azaleacolburn replied:
Ok, maybe make a note about what kind of types T work, if you have an idea that might be helpful to anyone else using this function.

HunterBarclay replied:
Honestly the issue is with horrid generics support for TypeScript. I think I'm going to dispense with it and leave it to the user.

azaleacolburn replied:
Ok ya that sounds good. Generic traits in TS when?

azaleacolburn reviewed (commented):
What's the plan for resolving this PR with preferences? Will there be another PR with an overlapping scope, or have they been completely bundled together?

HunterBarclay commented:
@Ewie21 Preferences will be merged in first, then this branch, then finally Ayla's branch will have all the glue for everything.

azaleacolburn commented:
> Notes for Reviewers
> 
> 1. checkselectednode, listrobots, etc.. are duplicated functions. Where should I put them
> 2. Can I optimized this in checkSelectedNode()
> 
> ```ts
> selectedRobot?.mirabufInstance?.parser.rigidNodes.forEach(rn => {
>             if (World.PhysicsSystem.GetBody(selectedRobot.mechanism.GetBodyByNodeId(rn.id)!).GetID() === body.GetID()) {
>                 bodyAttachmentRef.current = body
>                 returnValue = true
>             }
>         })
> ```

You could make it into a regular for loop and break, since all you'll do is set the same variables again. Also, could you just compare the ids directly, or are ids not consistent between Physics System and mechanism? I'm sure there are other things as well, but I haven't throughly read through the code yet.

HunterBarclay commented:
> You could make it into a regular for loop and break, since all you'll do is set the same variables again. Also, could you just compare the ids directly, or are ids not consistent between Physics System and mechanism? I'm sure there are other things as well, but I haven't throughly read through the code yet.

@Ewie21 This no longer exists.

HunterBarclay committed:
Merge branch 'dev' into dhruv/1728/intake-ejector-config

HunterBarclay commented:
Note: Rigid node highlighting can be implemented once #1023 is merged in.

HunterBarclay committed:
Renamed BodyAssociated to BodyAssociate, made requested changes.

HunterBarclay committed:
Adding unit tests

LucaHaverty commented:
I'm not sure when this was introduced, but opening the poker panel or settings modal throws this error. It might be anything that uses a slider.
```
TypeError: Cannot read properties of undefined (reading 'toLocaleString')
    at Slider (Slider.tsx:38:44)
    at renderWithHooks (react-dom.development.js:15486:18)
    at mountIndeterminateComponent (react-dom.development.js:20103:13)
    at beginWork (react-dom.development.js:21626:16)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:4164:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:4213:16)
    at invokeGuardedCallback (react-dom.development.js:4277:31)
    at beginWork$1 (react-dom.development.js:27490:7)
    at performUnitOfWork (react-dom.development.js:26596:12)
    at workLoopSync (react-dom.development.js:26505:5)
```

HunterBarclay commented:
@LucaHaverty  Oops, I'll take care of it

LucaHaverty commented on fission/src/ui/components/Slider.tsx:
```diff
@@ -12,16 +12,15 @@ type SliderProps = {
     label?: string
     min: number
     max: number
-    defaultValue: number
```
Perhaps this is the cause of the slider issues? Maybe due to a merge?

Dhruv-0-Arora replied:
@LucaHaverty could you look at that again and see if you can find errors with the fix
I didn't touch the ZoneConfigPanel because I don't want merge conflicts from your PR

LucaHaverty reviewed (commented):
Besides the slider things, I could not find any other issues, great work!

Dhruv-0-Arora committed:
Fixed poker slider error
Added changes to other files but ignored ZoneConfigPanel due to PR #1021 changing it

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1728/intake-ejector-config

Dhruv-0-Arora committed:
Formatting

LucaHaverty reviewed (approved):
Everything worked well, fantastic work!

HunterBarclay committed:
Fixing body associates

Dhruv-0-Arora commented on fission/src/mirabuf/IntakeSensorSceneObject.ts:
```diff
@@ -75,7 +75,8 @@ class IntakeSensorSceneObject extends SceneObject {
-                    const gpAssoc = World.PhysicsSystem.GetBodyAssociation<RigidNodeAssociate>(hitRes.data.mBodyID)
+                    const gpAssoc = <RigidNodeAssociate>World.PhysicsSystem.GetBodyAssociation(hitRes.data.mBodyID)
+                    // This works, however the check for game piece is doing two checks.
                     if (gpAssoc && gpAssoc.isGamePiece) {
```
could you do gpAssoc?.isGamePiece

HunterBarclay replied:
Yes I believe so

Dhruv-0-Arora commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -1,17 +1,22 @@
 import * as THREE from "three"
 import JOLT from "@/util/loading/JoltSyncLoader"
-import { LayerReserve } from "@/systems/physics/PhysicsSystem"
+import { BodyAssociate, JoltBodyIndexAndSequence, LayerReserve } from "@/systems/physics/PhysicsSystem"
```
```suggestion
import { BodyAssociate, LayerReserve } from "@/systems/physics/PhysicsSystem"
```

a-crowell replied:
Fixed

BrandonPacewic reviewed (commented):
@Dhruv-0-Arora Are you not the author of this PR? Or did @HunterBarclay take this one over?

Dhruv-0-Arora commented:
> @Dhruv-0-Arora Are you not the author of this PR? Or did @HunterBarclay take this one over?

Hunter is the owner for this one

a-crowell committed:
Removed unused var eslint complaint

a-crowell reviewed (approved):
Works great! Just removed an unused import.

BrandonPacewic reviewed (changes requested):
Want to make sure @Dhruv-0-Arora 's comments are looked at before merge.

> could you do gpAssoc?.isGamePiece
> 
> _Originally posted by @Dhruv-0-Arora in https://github.com/Autodesk/synthesis/pull/1017#discussion_r1678356692_
>
> @LucaHaverty could you look at that again and see if you can find errors with the fix
I didn't touch the ZoneConfigPanel because I don't want merge conflicts from your PR
> 
> _Originally posted by @Dhruv-0-Arora in https://github.com/Autodesk/synthesis/pull/1017#discussion_r1678271072_
>
> That is true it would be more optimized
>
> _Originally posted by @Dhruv-0-Arora in https://github.com/Autodesk/synthesis/pull/1017#discussion_r1671662056_

HunterBarclay committed:
made requested change
